samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,253 @@
+                    template_field = "TemplateURL" if resource_type == "AWS::CloudFormation::Stack" else "Location"
+                    if not self._compare_templates(
+                        resource_dict.get("Properties", {}).get(template_field),
+                        stack_resource_detail.get("StackResourceDetail", {}).get("PhysicalResourceId", ""),
+                    ):
+                        return False
+        except Exception:
+            LOG.debug("Template comparison with the cloud template failed, not skipping infra sync")
+            return False
+
+        return True
+
+    def _remove_unnecessary_fields(self, template_dict: Dict, linked_resources: List[str] = []) -> List[str]:
+        """
+        Fields skipped during template comparison because sync --code can handle the difference:
+        * CodeUri or ImageUri property of AWS::Serverless::Function
+        * ImageUri, S3Bucket, S3Key, S3ObjectVersion fields in Code property of AWS::Lambda::Function
+        * ContentUri property of AWS::Serverless::LayerVersion
+        * Content property of AWS::Lambda::LayerVersion
+        * DefinitionBody property of AWS::Serverless::Api
+        * BodyS3Location property of AWS::ApiGateway::RestApi
+        * DefinitionUri property of AWS::Serverless::HttpApi
+        * BodyS3Location property of AWS::ApiGatewayV2::Api
+        * DefinitionUri property of AWS::Serverless::StateMachine
+        * DefinitionS3Location property of AWS::StepFunctions::StateMachine
+        * SamResourceId in Metadata property of all resources
+
+        Parameters
+        ----------
+        template_dict: Dict
+            The unprocessed template dictionary
+        linked_resources: List[str]
+            The corresponding resources in the other template that got processed
+
+        Returns
+        -------
+        List[str]
+            The list of resource IDs that got changed during sanitization
+        """
+
+        resources = template_dict.get("Resources", {})
+        processed_resources: Set[str] = set()
+
+        for resource_logical_id in resources:
+            resource_dict = resources.get(resource_logical_id, {})
+            resource_type = resource_dict.get("Type")
+
+            if resource_type in SYNCABLE_RESOURCES:
+                processed_resources = self._remove_resource_field(
+                    resource_logical_id,
+                    resource_type,
+                    resource_dict,
+                    processed_resources,
+                    linked_resources,
+                )
+
+            # Remove SamResourceId metadata since this metadata does not affect any cloud behaviour
+            resource_dict.get("Metadata", {}).pop("SamResourceId", None)
+            if not resource_dict.get("Metadata"):
+                resource_dict.pop("Metadata", None)
+            LOG.debug(f"Sanitizing the Metadata for resource {resource_logical_id}")
```

Comment:
Do we know why `SamResourceId` is creating problems, can we find where it is been added and change our behaviour regarding to that? My worry is, if there are some other variables like that, we might hit the same problem later.

Reply:
```
        metadata_copied.pop(SAM_RESOURCE_ID_KEY, "")
        metadata_copied.pop(SAM_IS_NORMALIZED, "")
```
Seems like a value that's to be popped, but in our code base there are many hard coded references to "SamResourceId", haven't looked at which exact reference is causing the difference